{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 584,
   "metadata": {},
   "outputs": [],
   "source": [
    "# look for sequence inserted in transMap of AugustusPB back to human\n",
    "from tools.psl import *\n",
    "from tools.transcripts import *\n",
    "import os\n",
    "import pandas as pd\n",
    "from cat.consensus import *\n",
    "from collections import *\n",
    "from tools.fileOps import *\n",
    "from tools.sqlInterface import *\n",
    "from tools.nameConversions import strip_alignment_numbers\n",
    "data_dir = '/public/groups/cgl/cat/primates_evan/rhesus_reverse/work/transMap'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "alns = get_alignment_dict(os.path.join(data_dir, 'Human.filtered.psl'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "txs = get_gene_pred_dict(os.path.join(data_dir, 'Human.filtered.gp'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_txs = get_gene_pred_dict('/public/groups/cgl/cat/primates_evan/rhesus_reverse/work/reference/Rhesus.gp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = load_annotation('/public/groups/cgl/cat/primates_evan/rhesus_reverse/out/databases/Rhesus.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for 1-1 mappings\n",
    "unique = set(tm_eval_df[tm_eval_df.UnfilteredParalogy == ''].TranscriptId)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for novel *isoform* predictions\n",
    "novel = set(ref_df[(ref_df.TranscriptBiotype == 'unknown_likely_coding') & (ref_df.GeneBiotype != 'unknown_likely_coding')].TranscriptId)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4951\n"
     ]
    }
   ],
   "source": [
    "useful_ids = novel & unique\n",
    "print(len(useful_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_ref_chrom(ref_tx, left_pos, right_pos):\n",
    "    if ref_tx.strand == '+':\n",
    "        rhesus_right_chrom = ref_tx.mrna_coordinate_to_chromosome(right_pos)\n",
    "        rhesus_left_chrom = ref_tx.mrna_coordinate_to_chromosome(left_pos)\n",
    "    else:\n",
    "        rhesus_right_chrom = ref_tx.mrna_coordinate_to_chromosome(left_pos) + 1\n",
    "        rhesus_left_chrom = ref_tx.mrna_coordinate_to_chromosome(right_pos - 1)\n",
    "    assert rhesus_right_chrom > rhesus_left_chrom\n",
    "    return rhesus_left_chrom, rhesus_right_chrom\n",
    "\n",
    "\n",
    "def find_exons(ref_tx, rhesus_left_chrom, rhesus_right_chrom):\n",
    "    i = ChromosomeInterval(ref_tx.chromosome, rhesus_left_chrom, rhesus_right_chrom, ref_tx.strand)\n",
    "    intersections = []\n",
    "    for exon in ref_tx.exon_intervals:\n",
    "        if exon.overlap(i):\n",
    "            intersections.append(exon)\n",
    "    return intersections\n",
    "\n",
    "\n",
    "def convert_back_to_tx(ref_tx, exon):\n",
    "    # convert to tx coordinates\n",
    "    left_pos = ref_tx.chromosome_coordinate_to_mrna(exon.start)\n",
    "    right_pos = ref_tx.chromosome_coordinate_to_mrna(exon.stop - 1)\n",
    "    if right_pos < left_pos:\n",
    "        right_pos, left_pos = left_pos, right_pos\n",
    "    return left_pos, right_pos\n",
    "\n",
    "def find_indels(aln, tx, ref_tx, min_size=30):\n",
    "    # remember where we were last iteration\n",
    "    q_pos = 0\n",
    "    t_pos = 0\n",
    "    # iterate over block starts[i], q_starts[i + 1], t_starts[i + 1]\n",
    "    for block_size, q_start, t_start in zip(*[aln.block_sizes, aln.q_starts[1:], aln.t_starts[1:]]):\n",
    "        q_offset = q_start - block_size - q_pos\n",
    "        t_offset = t_start - block_size - t_pos\n",
    "        if q_offset != 0:  # query insertion -> insertion in target sequence\n",
    "            if tx.strand == '+':\n",
    "                left_pos = q_start - q_offset\n",
    "                right_pos = q_start\n",
    "                right_chrom = aln.query_coordinate_to_target(right_pos + 1)\n",
    "                left_chrom = aln.query_coordinate_to_target(left_pos - 1)\n",
    "            else:\n",
    "                left_pos = aln.q_size - q_start\n",
    "                right_pos = aln.q_size - q_start + q_offset\n",
    "                right_chrom = aln.query_coordinate_to_target(left_pos - 1)\n",
    "                left_chrom = aln.query_coordinate_to_target(right_pos + 1)\n",
    "            if left_chrom is None or right_chrom is None:\n",
    "                continue\n",
    "            assert left_chrom < right_chrom\n",
    "            assert left_pos < right_pos\n",
    "            size = right_pos - left_pos\n",
    "            if size >= min_size:\n",
    "                rhesus_left_chrom, rhesus_right_chrom = convert_ref_chrom(ref_tx, left_pos, right_pos)\n",
    "                for exon in find_exons(ref_tx, rhesus_left_chrom, rhesus_right_chrom):\n",
    "                    exon_left_pos, exon_right_pos = convert_back_to_tx(ref_tx, exon)\n",
    "                    yield tx.chromosome, left_chrom, right_chrom, tx.name, exon_left_pos, exon_right_pos, len(exon), exon.chromosome, exon.start, exon.stop\n",
    "        q_pos = q_start\n",
    "        t_pos = t_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 566,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = []\n",
    "# maintain a set to keep out duplicates (augPB/augCGP predicting the same novel exon more than once)\n",
    "seen_exons = set()\n",
    "for aln_id, tx in txs.items():\n",
    "    if strip_alignment_numbers(aln_id) not in useful_ids:\n",
    "        continue\n",
    "    aln = alns[aln_id]\n",
    "    ref_tx = ref_txs[strip_alignment_numbers(aln_id)]\n",
    "    for val in find_indels(aln, tx, ref_tx):\n",
    "        i = (val[-3], val[-2], val[-1])\n",
    "        if i in seen_exons:\n",
    "            continue\n",
    "        seen_exons.add(i)\n",
    "        r.append(val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "metadata": {},
   "outputs": [],
   "source": [
    "rdf = pd.DataFrame(r, columns=['chromosome', 'chrom_start', 'chrom_stop', 'aln_id', 'start', 'stop', 'size',\n",
    "                              'rhesus_chromosome', 'rhesus_chrom_start', 'rhesus_chrom_stop'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p bam_tmp\n",
    "rdf[['rhesus_chromosome', 'rhesus_chrom_start', 'rhesus_chrom_stop']].to_csv('bam_tmp/exons.bed', header=None, index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 598,
   "metadata": {},
   "outputs": [],
   "source": [
    "# grab CAT consensus transcripts that are not CGP/PB to filter out false positives\n",
    "rhesus_cat_df = pd.read_csv('/public/groups/cgl/cat/primates_evan/out/consensus_gene_set/Rhesus.gp_info', sep='\\t')\n",
    "to_keep = set(rhesus_cat_df[rhesus_cat_df.transcript_biotype != 'unknown_likely_coding'].transcript_id)\n",
    "with open('bam_tmp/rhesus_cat_exons.bed', 'w') as fh:\n",
    "    for tx in gene_pred_iterator('/public/groups/cgl/cat/primates_evan/out/consensus_gene_set/Rhesus.gp'):\n",
    "        if tx.name in to_keep:\n",
    "            for exon in tx.exon_intervals:\n",
    "                fh.write('\\t'.join(map(str, [tx.chromosome, exon.start, exon.stop])) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bedtools intersect -v -a bam_tmp/exons.bed -b bam_tmp/rhesus_cat_exons.bed > bam_tmp/exons_filtered_for_cat_consensus.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for IsoSeq support\n",
    "!samtools merge bam_tmp/merged.bam /public/groups/cgl/cat/primates_evan/out/assemblyHub/Rhesus/*bam\n",
    "!bedtools bamtobed -split -i bam_tmp/merged.bam > bam_tmp/merged.bed\n",
    "!bedSort bam_tmp/merged.bed bam_tmp/merged.bed\n",
    "!bedtools merge -i bam_tmp/merged.bed -c 1 -o count > bam_tmp/collapsed.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('bam_tmp/filtered.bed', 'w') as fh:\n",
    "    for l in open('bam_tmp/collapsed.bed'):\n",
    "        l = l.split()\n",
    "        if int(l[-1]) > 3:\n",
    "            fh.write('\\t'.join(l) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 600,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bedtools intersect -u -a bam_tmp/exons.bed -b bam_tmp/collapsed.bed > bam_tmp/exons_filtered_for_isoseq.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = rdf.copy()\n",
    "for bed, col in [['bam_tmp/exons_filtered_for_cat_consensus.bed', 'cat_consensus_filter'],\n",
    "                ['bam_tmp/exons_filtered_for_isoseq.bed', 'isoseq_filter']]:\n",
    "    tmp_df = pd.read_csv(bed, sep='\\t', header=None)\n",
    "    tmp_df.columns = ['rhesus_chromosome', 'rhesus_chrom_start', 'rhesus_chrom_stop']\n",
    "    tmp_df[col] = [True] * len(tmp_df)\n",
    "    m = m.merge(tmp_df, on=['rhesus_chromosome', 'rhesus_chrom_start', 'rhesus_chrom_stop'], how='left').drop_duplicates()\n",
    "m = m.fillna(False)\n",
    "# flip the consensus to match\n",
    "m['cat_consensus_filter'] = [not x for x in m['cat_consensus_filter']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 613,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add some reference info\n",
    "tmp_df = ref_df[['TranscriptId', 'GeneBiotype', 'GeneId', 'GeneName']]\n",
    "tmp_df.columns = ['aln_id', 'Gene Biotype', 'Gene ID', 'Gene Name']\n",
    "# strip the id from m\n",
    "m['aln_id'] = [strip_alignment_numbers(x) for x in m['aln_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 614,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = m.merge(tmp_df, on='aln_id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 616,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_map = OrderedDict([['Gene ID', 'Rhesus gene ID'],\n",
    "                       ['Gene Name', 'Gene name'],\n",
    "                       ['Gene Biotype', 'Gene type'],\n",
    "                       ['rhesus_chromosome', 'RheMac10 chromosome'],\n",
    "                       ['rhesus_chrom_start', 'RheMac10 start'],\n",
    "                       ['rhesus_chrom_stop', 'RheMac10 end'],\n",
    "                       ['chromosome', 'GRCh38 chromosome'],\n",
    "                       ['chrom_start', 'GRCh38 nearest chromosome start'],\n",
    "                       ['chrom_stop', 'GRCh38 nearest chromosome end'],\n",
    "                       ['size', 'Exon size'],\n",
    "                       ['cat_consensus_filter', 'Overlaps a CAT comparative transcript'],\n",
    "                       ['isoseq_filter', 'Has at least 3 supporting IsoSeq reads']])\n",
    "m = m.rename(columns=name_map)\n",
    "m = m[name_map.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 617,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = m.sort_values(['Overlaps a CAT comparative transcript',\n",
    "                  'RheMac10 chromosome',\n",
    "                  'RheMac10 start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 618,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "m.to_csv('rhesus_novel_exons.tsv', sep='\\t', index=None)\n",
    "m.to_excel('rhesus_novel_exons.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 619,
   "metadata": {},
   "outputs": [],
   "source": [
    "m[['RheMac10 chromosome', 'RheMac10 start', 'RheMac10 end']].to_csv('rhesus_unfiltered_novel_exons.bed', sep='\\t', header=None, index=None)\n",
    "m[m['Overlaps a CAT comparative transcript'] == False][['RheMac10 chromosome', 'RheMac10 start', 'RheMac10 end']].to_csv('rhesus_filtered_novel_exons.bed', sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 620,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rhesus gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>RheMac10 chromosome</th>\n",
       "      <th>RheMac10 start</th>\n",
       "      <th>RheMac10 end</th>\n",
       "      <th>GRCh38 chromosome</th>\n",
       "      <th>GRCh38 nearest chromosome start</th>\n",
       "      <th>GRCh38 nearest chromosome end</th>\n",
       "      <th>Exon size</th>\n",
       "      <th>Overlaps a CAT comparative transcript</th>\n",
       "      <th>Has at least 3 supporting IsoSeq reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rhesus_G0002154</td>\n",
       "      <td>S100A15A</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chr1</td>\n",
       "      <td>97658864</td>\n",
       "      <td>97659094</td>\n",
       "      <td>chr1</td>\n",
       "      <td>153396765</td>\n",
       "      <td>153400385</td>\n",
       "      <td>230</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rhesus_G0003082</td>\n",
       "      <td>HFM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>132859086</td>\n",
       "      <td>132859193</td>\n",
       "      <td>chr1</td>\n",
       "      <td>91277060</td>\n",
       "      <td>91324674</td>\n",
       "      <td>107</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Rhesus_G0003082</td>\n",
       "      <td>HFM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>132859288</td>\n",
       "      <td>132859336</td>\n",
       "      <td>chr1</td>\n",
       "      <td>91277060</td>\n",
       "      <td>91324674</td>\n",
       "      <td>48</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Rhesus_G0003082</td>\n",
       "      <td>HFM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>132863175</td>\n",
       "      <td>132863273</td>\n",
       "      <td>chr1</td>\n",
       "      <td>91277060</td>\n",
       "      <td>91324674</td>\n",
       "      <td>98</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Rhesus_G0003082</td>\n",
       "      <td>HFM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>132863355</td>\n",
       "      <td>132863487</td>\n",
       "      <td>chr1</td>\n",
       "      <td>91277060</td>\n",
       "      <td>91324674</td>\n",
       "      <td>132</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>Rhesus_G0056391</td>\n",
       "      <td>GPM6B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>13436359</td>\n",
       "      <td>13436425</td>\n",
       "      <td>chrX</td>\n",
       "      <td>13776257</td>\n",
       "      <td>13783364</td>\n",
       "      <td>66</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>Rhesus_G0056391</td>\n",
       "      <td>GPM6B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>13437441</td>\n",
       "      <td>13437515</td>\n",
       "      <td>chrX</td>\n",
       "      <td>13776257</td>\n",
       "      <td>13783364</td>\n",
       "      <td>74</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>Rhesus_G0056391</td>\n",
       "      <td>GPM6B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>13440180</td>\n",
       "      <td>13440352</td>\n",
       "      <td>chrX</td>\n",
       "      <td>13776257</td>\n",
       "      <td>13783364</td>\n",
       "      <td>172</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245</th>\n",
       "      <td>Rhesus_G0058144</td>\n",
       "      <td>MAP7D3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>132333022</td>\n",
       "      <td>132333651</td>\n",
       "      <td>chrX</td>\n",
       "      <td>136231799</td>\n",
       "      <td>136231801</td>\n",
       "      <td>629</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>Rhesus_G0058572</td>\n",
       "      <td>BCORP1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chrY</td>\n",
       "      <td>5272480</td>\n",
       "      <td>5272626</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19460097</td>\n",
       "      <td>19466455</td>\n",
       "      <td>146</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>251 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Rhesus gene ID Gene name                           Gene type  \\\n",
       "3    Rhesus_G0002154  S100A15A                processed_pseudogene   \n",
       "4    Rhesus_G0003082      HFM1                      protein_coding   \n",
       "5    Rhesus_G0003082      HFM1                      protein_coding   \n",
       "6    Rhesus_G0003082      HFM1                      protein_coding   \n",
       "7    Rhesus_G0003082      HFM1                      protein_coding   \n",
       "..               ...       ...                                 ...   \n",
       "239  Rhesus_G0056391     GPM6B                      protein_coding   \n",
       "240  Rhesus_G0056391     GPM6B                      protein_coding   \n",
       "241  Rhesus_G0056391     GPM6B                      protein_coding   \n",
       "245  Rhesus_G0058144    MAP7D3                      protein_coding   \n",
       "248  Rhesus_G0058572    BCORP1  transcribed_unprocessed_pseudogene   \n",
       "\n",
       "    RheMac10 chromosome  RheMac10 start  RheMac10 end GRCh38 chromosome  \\\n",
       "3                  chr1        97658864      97659094              chr1   \n",
       "4                  chr1       132859086     132859193              chr1   \n",
       "5                  chr1       132859288     132859336              chr1   \n",
       "6                  chr1       132863175     132863273              chr1   \n",
       "7                  chr1       132863355     132863487              chr1   \n",
       "..                  ...             ...           ...               ...   \n",
       "239                chrX        13436359      13436425              chrX   \n",
       "240                chrX        13437441      13437515              chrX   \n",
       "241                chrX        13440180      13440352              chrX   \n",
       "245                chrX       132333022     132333651              chrX   \n",
       "248                chrY         5272480       5272626              chrY   \n",
       "\n",
       "     GRCh38 nearest chromosome start  GRCh38 nearest chromosome end  \\\n",
       "3                          153396765                      153400385   \n",
       "4                           91277060                       91324674   \n",
       "5                           91277060                       91324674   \n",
       "6                           91277060                       91324674   \n",
       "7                           91277060                       91324674   \n",
       "..                               ...                            ...   \n",
       "239                         13776257                       13783364   \n",
       "240                         13776257                       13783364   \n",
       "241                         13776257                       13783364   \n",
       "245                        136231799                      136231801   \n",
       "248                         19460097                       19466455   \n",
       "\n",
       "     Exon size  Overlaps a CAT comparative transcript  \\\n",
       "3          230                                  False   \n",
       "4          107                                  False   \n",
       "5           48                                  False   \n",
       "6           98                                  False   \n",
       "7          132                                  False   \n",
       "..         ...                                    ...   \n",
       "239         66                                   True   \n",
       "240         74                                   True   \n",
       "241        172                                   True   \n",
       "245        629                                   True   \n",
       "248        146                                   True   \n",
       "\n",
       "     Has at least 3 supporting IsoSeq reads  \n",
       "3                                      True  \n",
       "4                                      True  \n",
       "5                                      True  \n",
       "6                                      True  \n",
       "7                                      True  \n",
       "..                                      ...  \n",
       "239                                    True  \n",
       "240                                    True  \n",
       "241                                    True  \n",
       "245                                    True  \n",
       "248                                    True  \n",
       "\n",
       "[251 rows x 12 columns]"
      ]
     },
     "execution_count": 620,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
