{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 584,
   "metadata": {},
   "outputs": [],
   "source": [
    "# look for sequence inserted in transMap of AugustusPB back to human\n",
    "from tools.psl import *\n",
    "from tools.transcripts import *\n",
    "import os\n",
    "import pandas as pd\n",
    "from cat.consensus import *\n",
    "from collections import *\n",
    "from tools.fileOps import *\n",
    "from tools.sqlInterface import *\n",
    "from tools.nameConversions import strip_alignment_numbers\n",
    "data_dir = '/public/groups/cgl/cat/primates_evan/v2/no-cupcake/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "alns = get_alignment_dict(os.path.join(data_dir, 'Human.filtered.psl'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "txs = get_gene_pred_dict(os.path.join(data_dir, 'Human.filtered.gp'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_txs = get_gene_pred_dict('/public/groups/cgl/cat/primates_evan/rhesus_reverse/work/reference/Rhesus.gp')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = load_annotation('/public/groups/cgl/cat/primates_evan/rhesus_reverse/out/databases/Rhesus.db')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for 1-1 mappings\n",
    "unique = set(tm_eval_df[tm_eval_df.UnfilteredParalogy == ''].TranscriptId)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for novel *isoform* predictions\n",
    "novel = set(ref_df[(ref_df.TranscriptBiotype == 'unknown_likely_coding') & (ref_df.GeneBiotype != 'unknown_likely_coding')].TranscriptId)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4951\n"
     ]
    }
   ],
   "source": [
    "useful_ids = novel & unique\n",
    "print(len(useful_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_ref_chrom(ref_tx, left_pos, right_pos):\n",
    "    if ref_tx.strand == '+':\n",
    "        rhesus_right_chrom = ref_tx.mrna_coordinate_to_chromosome(right_pos)\n",
    "        rhesus_left_chrom = ref_tx.mrna_coordinate_to_chromosome(left_pos)\n",
    "    else:\n",
    "        rhesus_right_chrom = ref_tx.mrna_coordinate_to_chromosome(left_pos) + 1\n",
    "        rhesus_left_chrom = ref_tx.mrna_coordinate_to_chromosome(right_pos - 1)\n",
    "    assert rhesus_right_chrom > rhesus_left_chrom\n",
    "    return rhesus_left_chrom, rhesus_right_chrom\n",
    "\n",
    "\n",
    "def find_exons(ref_tx, rhesus_left_chrom, rhesus_right_chrom):\n",
    "    i = ChromosomeInterval(ref_tx.chromosome, rhesus_left_chrom, rhesus_right_chrom, ref_tx.strand)\n",
    "    intersections = []\n",
    "    for exon in ref_tx.exon_intervals:\n",
    "        if exon.overlap(i):\n",
    "            intersections.append(exon)\n",
    "    return intersections\n",
    "\n",
    "\n",
    "def convert_back_to_tx(ref_tx, exon):\n",
    "    # convert to tx coordinates\n",
    "    left_pos = ref_tx.chromosome_coordinate_to_mrna(exon.start)\n",
    "    right_pos = ref_tx.chromosome_coordinate_to_mrna(exon.stop - 1)\n",
    "    if right_pos < left_pos:\n",
    "        right_pos, left_pos = left_pos, right_pos\n",
    "    return left_pos, right_pos\n",
    "\n",
    "def find_indels(aln, tx, ref_tx, min_size=30):\n",
    "    # remember where we were last iteration\n",
    "    q_pos = 0\n",
    "    t_pos = 0\n",
    "    # iterate over block starts[i], q_starts[i + 1], t_starts[i + 1]\n",
    "    for block_size, q_start, t_start in zip(*[aln.block_sizes, aln.q_starts[1:], aln.t_starts[1:]]):\n",
    "        q_offset = q_start - block_size - q_pos\n",
    "        t_offset = t_start - block_size - t_pos\n",
    "        if q_offset != 0:  # query insertion -> insertion in target sequence\n",
    "            if tx.strand == '+':\n",
    "                left_pos = q_start - q_offset\n",
    "                right_pos = q_start\n",
    "                right_chrom = aln.query_coordinate_to_target(right_pos + 1)\n",
    "                left_chrom = aln.query_coordinate_to_target(left_pos - 1)\n",
    "            else:\n",
    "                left_pos = aln.q_size - q_start\n",
    "                right_pos = aln.q_size - q_start + q_offset\n",
    "                right_chrom = aln.query_coordinate_to_target(left_pos - 1)\n",
    "                left_chrom = aln.query_coordinate_to_target(right_pos + 1)\n",
    "            if left_chrom is None or right_chrom is None:\n",
    "                continue\n",
    "            assert left_chrom < right_chrom\n",
    "            assert left_pos < right_pos\n",
    "            size = right_pos - left_pos\n",
    "            if size >= min_size:\n",
    "                rhesus_left_chrom, rhesus_right_chrom = convert_ref_chrom(ref_tx, left_pos, right_pos)\n",
    "                for exon in find_exons(ref_tx, rhesus_left_chrom, rhesus_right_chrom):\n",
    "                    exon_left_pos, exon_right_pos = convert_back_to_tx(ref_tx, exon)\n",
    "                    yield tx.chromosome, left_chrom, right_chrom, tx.name, exon_left_pos, exon_right_pos, len(exon), exon.chromosome, exon.start, exon.stop\n",
    "        q_pos = q_start\n",
    "        t_pos = t_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 566,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = []\n",
    "# maintain a set to keep out duplicates (augPB/augCGP predicting the same novel exon more than once)\n",
    "seen_exons = set()\n",
    "for aln_id, tx in txs.items():\n",
    "    if strip_alignment_numbers(aln_id) not in useful_ids:\n",
    "        continue\n",
    "    aln = alns[aln_id]\n",
    "    ref_tx = ref_txs[strip_alignment_numbers(aln_id)]\n",
    "    for val in find_indels(aln, tx, ref_tx):\n",
    "        i = (val[-3], val[-2], val[-1])\n",
    "        if i in seen_exons:\n",
    "            continue\n",
    "        seen_exons.add(i)\n",
    "        r.append(val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 567,
   "metadata": {},
   "outputs": [],
   "source": [
    "rdf = pd.DataFrame(r, columns=['chromosome', 'chrom_start', 'chrom_stop', 'aln_id', 'start', 'stop', 'size',\n",
    "                              'rhesus_chromosome', 'rhesus_chrom_start', 'rhesus_chrom_stop'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 568,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p bam_tmp\n",
    "rdf[['rhesus_chromosome', 'rhesus_chrom_start', 'rhesus_chrom_stop']].to_csv('bam_tmp/exons.bed', header=None, index=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 598,
   "metadata": {},
   "outputs": [],
   "source": [
    "# grab CAT consensus transcripts that are not CGP/PB to filter out false positives\n",
    "rhesus_cat_df = pd.read_csv('/public/groups/cgl/cat/primates_evan/out/consensus_gene_set/Rhesus.gp_info', sep='\\t')\n",
    "to_keep = set(rhesus_cat_df[rhesus_cat_df.transcript_biotype != 'unknown_likely_coding'].transcript_id)\n",
    "with open('bam_tmp/rhesus_cat_exons.bed', 'w') as fh:\n",
    "    for tx in gene_pred_iterator('/public/groups/cgl/cat/primates_evan/out/consensus_gene_set/Rhesus.gp'):\n",
    "        if tx.name in to_keep:\n",
    "            for exon in tx.exon_intervals:\n",
    "                fh.write('\\t'.join(map(str, [tx.chromosome, exon.start, exon.stop])) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 599,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bedtools intersect -v -a bam_tmp/exons.bed -b bam_tmp/rhesus_cat_exons.bed > bam_tmp/exons_filtered_for_cat_consensus.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter for IsoSeq support\n",
    "!samtools merge bam_tmp/merged.bam /public/groups/cgl/cat/primates_evan/out/assemblyHub/Rhesus/*bam\n",
    "!bedtools bamtobed -split -i bam_tmp/merged.bam > bam_tmp/merged.bed\n",
    "!bedSort bam_tmp/merged.bed bam_tmp/merged.bed\n",
    "!bedtools merge -i bam_tmp/merged.bed -c 1 -o count > bam_tmp/collapsed.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('bam_tmp/filtered.bed', 'w') as fh:\n",
    "    for l in open('bam_tmp/collapsed.bed'):\n",
    "        l = l.split()\n",
    "        if int(l[-1]) > 3:\n",
    "            fh.write('\\t'.join(l) + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 600,
   "metadata": {},
   "outputs": [],
   "source": [
    "!bedtools intersect -u -a bam_tmp/exons.bed -b bam_tmp/collapsed.bed > bam_tmp/exons_filtered_for_isoseq.bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = rdf.copy()\n",
    "for bed, col in [['bam_tmp/exons_filtered_for_cat_consensus.bed', 'cat_consensus_filter'],\n",
    "                ['bam_tmp/exons_filtered_for_isoseq.bed', 'isoseq_filter']]:\n",
    "    tmp_df = pd.read_csv(bed, sep='\\t', header=None)\n",
    "    tmp_df.columns = ['rhesus_chromosome', 'rhesus_chrom_start', 'rhesus_chrom_stop']\n",
    "    tmp_df[col] = [True] * len(tmp_df)\n",
    "    m = m.merge(tmp_df, on=['rhesus_chromosome', 'rhesus_chrom_start', 'rhesus_chrom_stop'], how='left').drop_duplicates()\n",
    "m = m.fillna(False)\n",
    "# flip the consensus to match\n",
    "m['cat_consensus_filter'] = [not x for x in m['cat_consensus_filter']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 613,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add some reference info\n",
    "tmp_df = ref_df[['TranscriptId', 'GeneBiotype', 'GeneId', 'GeneName']]\n",
    "tmp_df.columns = ['aln_id', 'Gene Biotype', 'Gene ID', 'Gene Name']\n",
    "# strip the id from m\n",
    "m['aln_id'] = [strip_alignment_numbers(x) for x in m['aln_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 614,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = m.merge(tmp_df, on='aln_id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 616,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_map = OrderedDict([['Gene ID', 'Rhesus gene ID'],\n",
    "                       ['Gene Name', 'Gene name'],\n",
    "                       ['Gene Biotype', 'Gene type'],\n",
    "                       ['rhesus_chromosome', 'RheMac10 chromosome'],\n",
    "                       ['rhesus_chrom_start', 'RheMac10 start'],\n",
    "                       ['rhesus_chrom_stop', 'RheMac10 end'],\n",
    "                       ['chromosome', 'GRCh38 chromosome'],\n",
    "                       ['chrom_start', 'GRCh38 nearest chromosome start'],\n",
    "                       ['chrom_stop', 'GRCh38 nearest chromosome end'],\n",
    "                       ['size', 'Exon size'],\n",
    "                       ['cat_consensus_filter', 'Overlaps a CAT comparative transcript'],\n",
    "                       ['isoseq_filter', 'Has at least 3 supporting IsoSeq reads']])\n",
    "m = m.rename(columns=name_map)\n",
    "m = m[name_map.values()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 617,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = m.sort_values(['Overlaps a CAT comparative transcript',\n",
    "                  'RheMac10 chromosome',\n",
    "                  'RheMac10 start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 618,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "m.to_csv('rhesus_novel_exons.tsv', sep='\\t', index=None)\n",
    "m.to_excel('rhesus_novel_exons.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 619,
   "metadata": {},
   "outputs": [],
   "source": [
    "m[['RheMac10 chromosome', 'RheMac10 start', 'RheMac10 end']].to_csv('rhesus_unfiltered_novel_exons.bed', sep='\\t', header=None, index=None)\n",
    "m[m['Overlaps a CAT comparative transcript'] == False][['RheMac10 chromosome', 'RheMac10 start', 'RheMac10 end']].to_csv('rhesus_filtered_novel_exons.bed', sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 622,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Rhesus gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>RheMac10 chromosome</th>\n",
       "      <th>RheMac10 start</th>\n",
       "      <th>RheMac10 end</th>\n",
       "      <th>GRCh38 chromosome</th>\n",
       "      <th>GRCh38 nearest chromosome start</th>\n",
       "      <th>GRCh38 nearest chromosome end</th>\n",
       "      <th>Exon size</th>\n",
       "      <th>Overlaps a CAT comparative transcript</th>\n",
       "      <th>Has at least 3 supporting IsoSeq reads</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>Rhesus_G0038279</td>\n",
       "      <td>LINC02538</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr4</td>\n",
       "      <td>3059481</td>\n",
       "      <td>3059586</td>\n",
       "      <td>chr6</td>\n",
       "      <td>167554683</td>\n",
       "      <td>167560118</td>\n",
       "      <td>105</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Rhesus_G0003082</td>\n",
       "      <td>HFM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1</td>\n",
       "      <td>132859288</td>\n",
       "      <td>132859336</td>\n",
       "      <td>chr1</td>\n",
       "      <td>91277060</td>\n",
       "      <td>91324674</td>\n",
       "      <td>48</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>Rhesus_G0019551</td>\n",
       "      <td>ATP8B5P</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chr15</td>\n",
       "      <td>48738233</td>\n",
       "      <td>48738417</td>\n",
       "      <td>chr9</td>\n",
       "      <td>35473770</td>\n",
       "      <td>35478242</td>\n",
       "      <td>184</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Rhesus_G0016025</td>\n",
       "      <td>LINC00301</td>\n",
       "      <td>transcribed_unitary_pseudogene</td>\n",
       "      <td>chr14</td>\n",
       "      <td>13576521</td>\n",
       "      <td>13576578</td>\n",
       "      <td>chr11</td>\n",
       "      <td>60621117</td>\n",
       "      <td>60628289</td>\n",
       "      <td>57</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Rhesus_G0021949</td>\n",
       "      <td>C17orf64</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr16</td>\n",
       "      <td>34981919</td>\n",
       "      <td>34982060</td>\n",
       "      <td>chr17</td>\n",
       "      <td>60429556</td>\n",
       "      <td>60431181</td>\n",
       "      <td>141</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>Rhesus_G0017928</td>\n",
       "      <td>C11orf88</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr14</td>\n",
       "      <td>104666432</td>\n",
       "      <td>104666507</td>\n",
       "      <td>chr11</td>\n",
       "      <td>111516109</td>\n",
       "      <td>111534412</td>\n",
       "      <td>75</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>Rhesus_G0031198</td>\n",
       "      <td>RYBP</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr2</td>\n",
       "      <td>128529808</td>\n",
       "      <td>128530055</td>\n",
       "      <td>chr3</td>\n",
       "      <td>72446619</td>\n",
       "      <td>72447094</td>\n",
       "      <td>247</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Rhesus_G0017781</td>\n",
       "      <td>AP000851.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>chr14</td>\n",
       "      <td>95880019</td>\n",
       "      <td>95880200</td>\n",
       "      <td>chr11</td>\n",
       "      <td>102607451</td>\n",
       "      <td>102611202</td>\n",
       "      <td>181</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>Rhesus_G0023317</td>\n",
       "      <td>SEPTIN9</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr16</td>\n",
       "      <td>73854118</td>\n",
       "      <td>73854722</td>\n",
       "      <td>chr17</td>\n",
       "      <td>77481550</td>\n",
       "      <td>77481611</td>\n",
       "      <td>604</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Rhesus_G0011127</td>\n",
       "      <td>POLE</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr11</td>\n",
       "      <td>132479735</td>\n",
       "      <td>132480110</td>\n",
       "      <td>chr12</td>\n",
       "      <td>132647728</td>\n",
       "      <td>132647730</td>\n",
       "      <td>375</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>177</th>\n",
       "      <td>Rhesus_G0042031</td>\n",
       "      <td>GPAT3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr5</td>\n",
       "      <td>50170049</td>\n",
       "      <td>50170101</td>\n",
       "      <td>chr4</td>\n",
       "      <td>83486850</td>\n",
       "      <td>83536504</td>\n",
       "      <td>52</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Rhesus_G0021005</td>\n",
       "      <td>SHBG</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr16</td>\n",
       "      <td>7408675</td>\n",
       "      <td>7408939</td>\n",
       "      <td>chr17</td>\n",
       "      <td>7626224</td>\n",
       "      <td>7627985</td>\n",
       "      <td>264</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rhesus_G0002154</td>\n",
       "      <td>S100A15A</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>chr1</td>\n",
       "      <td>97658864</td>\n",
       "      <td>97659094</td>\n",
       "      <td>chr1</td>\n",
       "      <td>153396765</td>\n",
       "      <td>153400385</td>\n",
       "      <td>230</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>Rhesus_G0047383</td>\n",
       "      <td>TGM7</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr7</td>\n",
       "      <td>19778285</td>\n",
       "      <td>19778342</td>\n",
       "      <td>chr15</td>\n",
       "      <td>43279949</td>\n",
       "      <td>43281843</td>\n",
       "      <td>57</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>Rhesus_G0019551</td>\n",
       "      <td>ATP8B5P</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chr15</td>\n",
       "      <td>48739505</td>\n",
       "      <td>48739611</td>\n",
       "      <td>chr9</td>\n",
       "      <td>35473770</td>\n",
       "      <td>35478242</td>\n",
       "      <td>106</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Rhesus_G0010451</td>\n",
       "      <td>AC126177.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>chr11</td>\n",
       "      <td>107516992</td>\n",
       "      <td>107517283</td>\n",
       "      <td>chr12</td>\n",
       "      <td>107873019</td>\n",
       "      <td>107882812</td>\n",
       "      <td>291</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>203</th>\n",
       "      <td>Rhesus_G0052820</td>\n",
       "      <td>ZFPM2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr8</td>\n",
       "      <td>105908140</td>\n",
       "      <td>105908254</td>\n",
       "      <td>chr8</td>\n",
       "      <td>105561480</td>\n",
       "      <td>105788718</td>\n",
       "      <td>114</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>Rhesus_G0028546</td>\n",
       "      <td>AC011487.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>chr19</td>\n",
       "      <td>53362139</td>\n",
       "      <td>53362295</td>\n",
       "      <td>chr19</td>\n",
       "      <td>53608445</td>\n",
       "      <td>53616926</td>\n",
       "      <td>156</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242</th>\n",
       "      <td>Rhesus_G0056458</td>\n",
       "      <td>BEND2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chrX</td>\n",
       "      <td>17957024</td>\n",
       "      <td>17957150</td>\n",
       "      <td>chrX</td>\n",
       "      <td>18195440</td>\n",
       "      <td>18201814</td>\n",
       "      <td>126</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Rhesus_G0005338</td>\n",
       "      <td>TRABD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr10</td>\n",
       "      <td>1291961</td>\n",
       "      <td>1292068</td>\n",
       "      <td>chr22</td>\n",
       "      <td>50189629</td>\n",
       "      <td>50189715</td>\n",
       "      <td>107</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Rhesus gene ID   Gene name                           Gene type  \\\n",
       "158  Rhesus_G0038279   LINC02538                              lncRNA   \n",
       "5    Rhesus_G0003082        HFM1                      protein_coding   \n",
       "68   Rhesus_G0019551     ATP8B5P  transcribed_unprocessed_pseudogene   \n",
       "51   Rhesus_G0016025   LINC00301      transcribed_unitary_pseudogene   \n",
       "83   Rhesus_G0021949    C17orf64                      protein_coding   \n",
       "60   Rhesus_G0017928    C11orf88                      protein_coding   \n",
       "120  Rhesus_G0031198        RYBP                      protein_coding   \n",
       "59   Rhesus_G0017781  AP000851.1                              lncRNA   \n",
       "86   Rhesus_G0023317     SEPTIN9                      protein_coding   \n",
       "36   Rhesus_G0011127        POLE                      protein_coding   \n",
       "177  Rhesus_G0042031       GPAT3                      protein_coding   \n",
       "80   Rhesus_G0021005        SHBG                      protein_coding   \n",
       "3    Rhesus_G0002154    S100A15A                processed_pseudogene   \n",
       "193  Rhesus_G0047383        TGM7                      protein_coding   \n",
       "70   Rhesus_G0019551     ATP8B5P  transcribed_unprocessed_pseudogene   \n",
       "34   Rhesus_G0010451  AC126177.5              unprocessed_pseudogene   \n",
       "203  Rhesus_G0052820       ZFPM2                      protein_coding   \n",
       "112  Rhesus_G0028546  AC011487.2  transcribed_unprocessed_pseudogene   \n",
       "242  Rhesus_G0056458       BEND2                      protein_coding   \n",
       "20   Rhesus_G0005338       TRABD                      protein_coding   \n",
       "\n",
       "    RheMac10 chromosome  RheMac10 start  RheMac10 end GRCh38 chromosome  \\\n",
       "158                chr4         3059481       3059586              chr6   \n",
       "5                  chr1       132859288     132859336              chr1   \n",
       "68                chr15        48738233      48738417              chr9   \n",
       "51                chr14        13576521      13576578             chr11   \n",
       "83                chr16        34981919      34982060             chr17   \n",
       "60                chr14       104666432     104666507             chr11   \n",
       "120                chr2       128529808     128530055              chr3   \n",
       "59                chr14        95880019      95880200             chr11   \n",
       "86                chr16        73854118      73854722             chr17   \n",
       "36                chr11       132479735     132480110             chr12   \n",
       "177                chr5        50170049      50170101              chr4   \n",
       "80                chr16         7408675       7408939             chr17   \n",
       "3                  chr1        97658864      97659094              chr1   \n",
       "193                chr7        19778285      19778342             chr15   \n",
       "70                chr15        48739505      48739611              chr9   \n",
       "34                chr11       107516992     107517283             chr12   \n",
       "203                chr8       105908140     105908254              chr8   \n",
       "112               chr19        53362139      53362295             chr19   \n",
       "242                chrX        17957024      17957150              chrX   \n",
       "20                chr10         1291961       1292068             chr22   \n",
       "\n",
       "     GRCh38 nearest chromosome start  GRCh38 nearest chromosome end  \\\n",
       "158                        167554683                      167560118   \n",
       "5                           91277060                       91324674   \n",
       "68                          35473770                       35478242   \n",
       "51                          60621117                       60628289   \n",
       "83                          60429556                       60431181   \n",
       "60                         111516109                      111534412   \n",
       "120                         72446619                       72447094   \n",
       "59                         102607451                      102611202   \n",
       "86                          77481550                       77481611   \n",
       "36                         132647728                      132647730   \n",
       "177                         83486850                       83536504   \n",
       "80                           7626224                        7627985   \n",
       "3                          153396765                      153400385   \n",
       "193                         43279949                       43281843   \n",
       "70                          35473770                       35478242   \n",
       "34                         107873019                      107882812   \n",
       "203                        105561480                      105788718   \n",
       "112                         53608445                       53616926   \n",
       "242                         18195440                       18201814   \n",
       "20                          50189629                       50189715   \n",
       "\n",
       "     Exon size  Overlaps a CAT comparative transcript  \\\n",
       "158        105                                  False   \n",
       "5           48                                  False   \n",
       "68         184                                  False   \n",
       "51          57                                  False   \n",
       "83         141                                  False   \n",
       "60          75                                  False   \n",
       "120        247                                  False   \n",
       "59         181                                  False   \n",
       "86         604                                  False   \n",
       "36         375                                  False   \n",
       "177         52                                  False   \n",
       "80         264                                  False   \n",
       "3          230                                  False   \n",
       "193         57                                  False   \n",
       "70         106                                  False   \n",
       "34         291                                  False   \n",
       "203        114                                  False   \n",
       "112        156                                  False   \n",
       "242        126                                  False   \n",
       "20         107                                  False   \n",
       "\n",
       "     Has at least 3 supporting IsoSeq reads  \n",
       "158                                    True  \n",
       "5                                      True  \n",
       "68                                     True  \n",
       "51                                     True  \n",
       "83                                     True  \n",
       "60                                     True  \n",
       "120                                    True  \n",
       "59                                     True  \n",
       "86                                     True  \n",
       "36                                     True  \n",
       "177                                    True  \n",
       "80                                     True  \n",
       "3                                      True  \n",
       "193                                    True  \n",
       "70                                     True  \n",
       "34                                     True  \n",
       "203                                    True  \n",
       "112                                    True  \n",
       "242                                    True  \n",
       "20                                     True  "
      ]
     },
     "execution_count": 622,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m[m['Overlaps a CAT comparative transcript'] == False].sample(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
